torch_xla_py/xla_dist.py

```diff
@@ -78,38 +114,158 @@ def validate(self):
         raise RuntimeError(
-            'All client_workers must have the same machine_type, got: {}'.format(
-                client_machine_types))
+            'All client_workers must have the same machine_type, got: {}'
+            .format(client_machine_types))
 
     if self._check_service_machine_type:
       server_machine_types = {
-          worker._machine_type for worker in self._service_workers}
+          worker._machine_type for worker in self._service_workers
+      }
       if len(server_machine_types) != 1:
         raise RuntimeError(
-            'All service_workers must have the same machine_type, got: {}'.format(
-                server_machine_types))
+            'All service_workers must have the same machine_type, got: {}'
+            .format(server_machine_types))
 
 
 class ClusterResolver(object):
   """Cluster Resolver for Client VM and Cloud TPU mesh."""
 
-  def __init__(self,
-               tpus,
-               vms=None,
-               zone=None,
-               project=None):
+  @staticmethod
+  def _get_instance_metadata(metadata):
+    response = requests.get(
+        '{}/computeMetadata/v1/{}'.format(_GCE_METADATA_ENDPOINT, metadata),
+        headers={'Metadata-Flavor': 'Google'})
+    return response.content.decode('utf-8')
+
+  def __init__(self, tpus, vms=None, zone=None, project=None):
     """Creates a new ClusterResolver object."""
 
-    assert tpus, "TPU name list must not be empty."
+    if not isinstance(tpus, list) or len(tpus) == 0:
+      raise ValueError('tpus must be a non-empty list')
+    if vms is not None:
+      if not isinstance(vms, list) or len(vms) == 0:
+        raise ValueError('vms must be a non-empty list if provided')
 
     self._tpus = tpus
     self._vms = vms
     self._zone = zone
     self._project = project
 
+    self._credentials = GoogleCredentials.get_application_default()
+    self._tpu_service = discovery.build(
+        'tpu', 'v1', credentials=self._credentials, cache_discovery=False)
+    self._compute_service = discovery.build(
+        'compute', 'v1', credentials=self._credentials, cache_discovery=False)
+
+    if project is None:
+      self._project = self._get_instance_metadata('project/project-id')
+    if zone is None:
+      zone_path = self._get_instance_metadata('instance/zone')
+      self._zone = os.path.split(zone_path)[1]
+    self._vm_master = self._get_instance_metadata('instance/name')
+
+  def _get_instance_group(self):
+    """Gets the instance group that the current VM belongs to."""
+    resp = self._compute_service.instances().get(
+        project=self._project,
+        zone=self._zone,
+        instance=self._vm_master,
+        fields='metadata').execute()
+
+    if 'metadata' in resp and 'items' in resp['metadata']:
+      for item in resp['metadata']['items']:
+        if (item['key'] == 'created-by' and
+            'instanceGroupManagers' in item['value']):
+          return os.path.split(item['value'])[1]
```

Comment:
I know I said os.path, but if this is referring to urls, then we should keep this as `.split('/')`. Sorry

Reply:
No worries, yes it's a URL path to the resource. Un-done.